src/createServer.ts

```diff
@@ -1,91 +1,84 @@
-import {
-  RequestPackageFunction,
-  RequestPackageOptions,
-  WantedDependency,
-} from '@pnpm/package-requester'
-import JsonSocket = require('json-socket')
-import net = require('net')
+import http = require('http')
+import {IncomingMessage, Server, ServerResponse} from 'http'
 import {StoreController} from 'package-store'
 
 export default function (
   store: StoreController,
-  opts: object,
+  opts: {
+    path?: string,
+    port?: number,
+    hostname?: string,
+  },
 ) {
-  const server = net.createServer()
-  server.listen(opts)
-  server.on('connection', (socket) => {
-    const jsonSocket = new JsonSocket(socket)
-    const requestPackage = requestPackageWithCtx.bind(null, {jsonSocket, store})
+  const manifestPromises = {}
+  const filesPromises = {}
 
-    jsonSocket.on('message', async (message) => {
-      switch (message.action) {
-        case 'requestPackage': {
-          await requestPackage(message.msgId, message.args[0], message.args[1])
-          return
-        }
-        case 'prune': {
-          await store.prune()
-          return
-        }
-        case 'updateConnections': {
-          await store.updateConnections(message.args[0], message.args[1])
-          return
+  const server = http.createServer(async (req: IncomingMessage, res: ServerResponse) => {
+    if (req.method !== 'POST') {
+      res.statusCode = 503
+      res.end(JSON.stringify(`Only POST is allowed, received ${req.method}`))
+      return
+    }
+
+    let body: any = '' // tslint:disable-line
+    req.on('data', (data) => {
+      body += data
+    })
+    req.on('end', async () => {
+      try {
+        if (body.length > 0) {
+          body = JSON.parse(body)
+        } else {
+          body = {}
         }
-        case 'saveState': {
-          await store.saveState()
-          return
+
+        switch (req.url) {
```

Comment:
the switch statement should be probably executed early, because if an unknown route is requested, there's no need to read the data

Reply:
The switch cannot be executed early because we don't have the body until the request has been received, and I don't see any reason to add complexity to wait for the body since we always need the body or it is empty so it doesn't add overhead.

Reply:
isn't the URL known from the very start? The URL can be validated before collecting the body

Reply:
and once we validate the URL? We've to wait for the body... so I am not really sure why we should validate the URL first